File: my-folder/0680-valid-palindrome-ii/solution.java
class Solution {
    public boolean validPalindrome(String s) {
        if (s == null || s.length() <= 1) return true;
        
        int start = 0;
        int end = s.length() - 1;
        
        while (start < end) {
            if (s.charAt(start) != s.charAt(end)) {
                return isPalindrome(s, start+1, end) || isPalindrome(s, start, end-1);
            }
            start++;
            end--;
        }
        return true;
    }
    
    private boolean isPalindrome(String s, int i, int j) {
        while (i < j) {
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

File: my-folder/0680-valid-palindrome-ii/solution.py
class Solution(object):
    def validPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        left = 0
        right = len(s)-1
        while left < right:
            if s[left]!=s[right]:
                if s[left+1: right+1]==s[right: left: -1] or \
                s[left: right]==s[-len(s)+right-1: -len(s)+left-1:-1]:
                    return True
                else:
                    return False
            else:
                left+=1
                right-=1
        return True
        
